backends/tc/ebpfCodeGen.cpp

```diff
@@ -0,0 +1,1481 @@
+/*
+Copyright (C) 2023 Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions
+and limitations under the License.
+*/
+
+#include "ebpfCodeGen.h"
+
+namespace TC {
+
+// =====================PNAEbpfGenerator=============================
+void PNAEbpfGenerator::emitPNAIncludes(EBPF::CodeBuilder *builder) const {
+    builder->appendLine("#include <stdbool.h>");
+    builder->appendLine("#include <linux/if_ether.h>");
+    builder->appendLine("#include \"pna.h\"");
+}
+
+void PNAEbpfGenerator::emitPreamble(EBPF::CodeBuilder *builder) const {
+    emitCommonPreamble(builder);
+    builder->newline();
+}
+
+void PNAEbpfGenerator::emitCommonPreamble(EBPF::CodeBuilder *builder) const {
+    builder->newline();
+    builder->appendLine("#define EBPF_MASK(t, w) ((((t)(1)) << (w)) - (t)1)");
+    builder->appendLine("#define BYTES(w) ((w) / 8)");
+    builder->appendLine(
+        "#define write_partial(a, w, s, v) do { *((u8*)a) = ((*((u8*)a)) "
+        "& ~(EBPF_MASK(u8, w) << s)) | (v << s) ; } while (0)");
+    builder->appendLine(
+        "#define write_byte(base, offset, v) do { "
+        "*(u8*)((base) + (offset)) = (v); "
+        "} while (0)");
+    builder->target->emitPreamble(builder);
+}
+
+void PNAEbpfGenerator::emitInternalStructures(EBPF::CodeBuilder *builder) const {
+    builder->appendLine(
+        "struct internal_metadata {\n"
+        "    __u16 pkt_ether_type;\n"
+        "} __attribute__((aligned(4)));");
+    builder->newline();
+}
+
+/* Generate headers and structs in p4 prog */
+void PNAEbpfGenerator::emitTypes(EBPF::CodeBuilder *builder) const {
+    PNAErrorCodesGen errorGen(builder);
+    pipeline->program->apply(errorGen);
+
+    for (auto type : ebpfTypes) {
+        type->emit(builder);
+    }
+
+    pipeline->parser->emitTypes(builder);
+    pipeline->control->emitTableTypes(builder);
+    pipeline->deparser->emitTypes(builder);
+    builder->newline();
+}
+
+void PNAEbpfGenerator::emitGlobalHeadersMetadata(EBPF::CodeBuilder *builder) const {
```

Comment:
There is a lot of copy-paste from the eBPF backend. Can we derive from `PSAEbpfGenerator` and override required methods? If there are some modifications to `PSAEbpfGenerator` needed, I'm open to apply them.

Reply:
There were some functions which needed to be overridden but were not defined as virtual in PSAEbpfGenerator class. Hence replicated similar code and did not derive from PSAEbpfGenerator class.

Reply:
updated the code to have common abstract class